File: src/main/java/com/tomatokey/framework/configuration/mvc/auth/Authorize.java
package com.tomatokey.framework.configuration.mvc.auth;

import com.tomatokey.architecture.layer_03_domain.userrole.UserRoleType;
import org.springframework.core.annotation.AliasFor;
import org.springframework.web.bind.annotation.RestController;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * {@link RestController}アノテーションが付与されたクラスまたは
 * そのクラス内のメソッドに付与すると認証の対象にすることができます
 */
@Target({ElementType.TYPE, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface Authorize {

    @AliasFor("roles")
    UserRoleType[] value() default {};

    /**
     * アクセスを許可する権限を設定します
     *
     * @return
     */
    @AliasFor("value")
    UserRoleType[] roles() default {};

}

File: src/main/kotlin/com/prototype/framework/configuration/mvc/auth/Authorize.kt
package com.prototype.framework.configuration.mvc.auth

import com.prototype.architecture.layer_03_domain.user.role.UserRoleType
import org.springframework.core.annotation.AliasFor

/**
 * [RestController]アノテーションが付与されたクラスまたは
 * そのクラス内のメソッドに付与すると認証の対象にすることができます
 */
@Target(
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.CLASS,
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER
)
@Retention(AnnotationRetention.RUNTIME)
annotation class Authorize(
    @get:AliasFor("roles")
    vararg val value: UserRoleType = [],
    @get:AliasFor("value")
    val roles: Array<UserRoleType> = []
)